game-app/game-core/src/main/java/games/strategy/engine/data/GameMap.java

```diff
@@ -396,7 +396,7 @@ private int getDistance(
 
     final Set<Territory> newFrontier =
         frontier.stream()
-            .flatMap(f -> connections.get(f).stream().filter(t -> routeCond.test(f, t)))
+            .flatMap(f -> getNeighbors(f).stream().filter(t -> routeCond.test(f, t)))
```

Comment:
`getNeighbors` is 'fancy' with a lot of overrides so it can accept a filtering condition.

Presumably this could be something like `.flatMap(frontier -> getNeighbors(frontier, 1, (start, end) -> routeCond.test(start, end));`

Though... I'm looking at `getNeighbors` and feel like it is eagerly DRY, would be better more explicit. In this case avoiding `getNeighbors` would be a good move. In this case, perhaps it would be better to do some simple java:
``` Optional.ofNullable(connections.get(f)).orElse(Set::of).stream....```

This is not a critical change, just pointing out that it's a bit odd for `getNeighbors` to be super overloaded to do route filtering, and then here use a default empty route filter in `getNeighbors` to then do route filtering (IMO, having the route filtering be here and not in the method is more clear, FWIW).

Reply:
The one-arg version of getNeighbors is just a wrapper over connections.get with a null check that provides a more useful error message. So I swapped all the connections.get calls to it.

Agree that this could be improved further.

Reply:
As to orElse(), I explicitly omitted it since I believe only time it should happen is a coding error - e.g. passing null or a territory from a different game object.

Reply:
Anyway, in this case you're right that it was pretty trivial to inline the filtering into this call site. Done.

Reply:
Re-reading I see you actually suggested to not use that for DRY. I can revert back if you prefer.

But expanding on my previous point, the getNeighbors() call was just calling the very base version that's just a wrapper over connections.get().

```
  public Set<Territory> getNeighbors(final Territory territory) {
    // ok since all entries in connections are already unmodifiable
    final Set<Territory> neighbors = connections.get(territory);
    if (neighbors == null) {
      throw new IllegalArgumentException("No neighbors for: " + territory);
    }
    return neighbors;
  }
```

Doing `orElse()` makes the code inconsistent - where some places handle the null and others throw an error. By calling this function and it being the only place that does `connections.get()` we make all the logic consistent with regards to expectations and handling of that case.

As an aside, I'm not very keen on the fact that most of GameMap is filled with different variants of getNeighbors(), getRoute(), getDistance(), etc, that all do mostly the same thing, with a complicated control flow and inefficiency of creating lots of intermediate collections.

IMHO, ideally we could extract everything except the base version above that just returns immediate neighbors - and use a helper class for all the other variants. I think I started towards that at some point with the BFS class, but there's quite a bit work to be done to make it handle the different cases and to change call sites. :\

As for this change, let me know what you prefer. For the NPE fix, we could just revert all the changes in GameMap - they're not strictly needed. But I thought they were worthwhile to make to handling of connections.get() consistent across the different places that call it. Another option is to remove the extra check in the base getNeighbors() impl and keep everyone using connections.get() directly. Downside is the error would just be a plain NPE which wouldn't be as obvious the cause of.